tests/integration/test_permissions_drop_replica/test.py

```diff
@@ -32,34 +53,55 @@ def start_cluster():
         cluster.shutdown()
 
 
+def check_exists(zk, path):
+    zk.sync(path)
+    return zk.exists(path)
+
+
 def test_drop_permissions(start_cluster):
-    node1.query("DROP DATABASE IF EXISTS r")
+    fill_nodes([node1, node2], 1)
     node1.query(
-        f"CREATE DATABASE r ENGINE=Replicated('/clickhouse/databases/r', '{{shard}}', '{{replica}}')"
+        "INSERT INTO test.test_table SELECT number, toString(number) FROM numbers(100)"
     )
-    node1.query(
-        "CREATE TABLE r.t1 (x UInt8, y String) ENGINE=ReplicatedMergeTree ORDER BY x"
+
+    # check metadata for replica node2 in keeper
+    # this the node replica that we are going to drop later on
+    zk = cluster.get_kazoo_client("zoo1")
+    # check that the path from zk exists for the replica node2
+
+    exists_before = check_exists(
+        zk, "/clickhouse/tables/test/1/replicated/test_table/replicas/node2"
+    )
+    assert exists_before != None
+
+    # we won't be able to drop an active replica
+    assert (
+        "DB::Exception: Can't drop replica: node2, because it's active."
+        in node1.query_and_get_error("SYSTEM DROP REPLICA 'node2'")
+    )
+
+    # stop the server on replica node2 and assume it's gone forever
+    # this will allow us to drop this replica
+    node2.stop_clickhouse()
+
+    # drop replica node2 from replica node1 using user without privileges
+    got_error = node1.query_and_get_error(
+        "SYSTEM DROP REPLICA 'node2'", user="test_user_xnhds"
     )
-    # create user without any permissions
-    node1.query("CREATE USER foo;")
-    node1.query("REVOKE ALL ON *.* FROM foo;")
-    # try dropping replica using this user without any permissions
-    got_error = node1.query_and_get_error("SYSTEM DROP REPLICA 'node2'", user="foo")
     # this operation should not fail silently
     assert (
         "DB::Exception: Access denied for SYSTEM DROP REPLICA. Not enough permissions to drop these databases:"
         in got_error
     )
-    # assert that the replica still exists
-    assert (
-        node1.query("SELECT host_name FROM system.clusters WHERE replica_num=1") != ""
-    )
-    # now query using default user (should have necessary permissions)
-    node1.query("SYSTEM DROP REPLICA 'node2'", user="default")
-    # assert that the replica was dropped successfully
-    assert (
-        node1.query("SELECT host_name FROM system.clusters WHERE replica_num=3").strip()
-        == ""
+
+    # drop replica node2 from replica node1 but using user with privileges
+    # this will remove all replicated table metadata belonging to node2 from keeper
+    node1.query("SYSTEM DROP REPLICA 'node2'")
+
+    # check that the metadata for replica node2 was removed from keeper
+    exists_after = check_exists(
+        zk, " /clickhouse/tables/test/1/replicated/test_table/replicas/node2"
     )
-    node1.query("DROP USER foo;")
```

Comment:
Let's leave cleanup stage in the end?

Reply:
It's now being handled in `fill_nodes()` (similar to other tests like `test_drop_replica` etc). I realized that this setup works more reliably when the tests are run more than once and in parallel.

Reply:
I do not agree. It is much better to expect that everything is cleaned up on test teardown and expect nothing to be on the cluster where we start a test. If we have some database/user on test start that means something is broken in this case.

Reply:
I've fixed to move the cleanup to the end.